How it works revision (#648)
dfinity/portal: how-it-works/1_protocol/01_peer-to-peer-p2p.subpage.md

Artifacts are the network messages that are to be broadcast in the subnet, such as ingress messages to canister smart contracts submitted by users or protocol-originating messages such as signature shares computed by the consensus layer of subnet nodes.
P2P realizes a peer-to-peer broadcast network that guarantees the secure eventual broadcast delivery of an artifact.
The P2P layer thereby is the communication fabric for the IC protocol stack and is used by the consensus layer, the next layer in the stack above it, to broadcast artifacts to the nodes in the subnet.

## Gossip Protocol

The P2P layer builds upon the basic principle of *gossip*.
Gossip in communication networks works along the same basic principle as gossip among people: A node in the subnet is connected with a subset of the other nodes of the subnet – its *peers*.
Whenever a node receives a message from a peer or creates one itself as part of the IC protocol, it gossips this message to all its peers.
By every node in the network doing exactly this, every artifact *eventually* propagates through the whole subnet.

> ## Asynchronous Communication Networks

Reviewer: Why does this come up here? Reliable broadcast doesn't really seem harder in the asynchronous setting, but consensus certainly is. Shall we maybe move it there?

Reviewer 2: Agreed

This important property of an *asynchronous communication network*, which reflects the real world of networks, is crucial and the IC consensus protocol is built with this assumption.
Clearly, assuming an asynchronous network makes protocols building on top more complex than they would be when assuming a synchronous network.

## Adverts

Ensuring the eventual delivery of broadcasts to all nodes in the subnet is achieved by a node providing a received or protocol-generated message to all its connected peers.
However, doing so in a naïve way by simply forwarding the message to all peers would result in nodes receiving as many copies of the same message as the node has peers, which unnecessarily consumes networking bandwidth for transmitting duplicate messages and reduces the achieveable throughput of the subnet.

This duplication of delivered artifacts in the naïve approach is mitigated by nodes sending *adverts* for artifacts to their peers instead of sending the artifacts themselves.
An advert specifies its corresponding artifact, but is a small message only containing the hash of the artifact to unambiguously refer to it and some additional metadata.

> Only adverts for artifacts are sent (pushed) to all its peers by a node.

Reviewer 2: ```suggestion
A node only pushes adverts for artifacts to its peers.
```

> A node receiving adverts then requests each artifact only once from one of its peers that has advertised the artifact through an advert.

Reviewer 2: ```suggestion
After receiving an advert, a node may request the corresponding artifact from one or more of its peers who sent it an advert for that artifact.
```

> For artifacts that are larger than the advert size, this saves bandwidth on the network connections between the peers – the savings grow with increasing size of the artifacts.

Reviewer 2: I'd remove this sentence. We indeed don't have a mechanism to send small artifacts immediately, and this sentence just raises the question why without answering. The answer is that it's buried somewhere in the backlog and the gain is not big enough to invest the required time :)

## Prioritization of Artifacts

The P2P layer allows the prioritization of artifacts such that the more crucial artifacts are broadcast throughout the subnet nodes more quickly than the others.
Prioritizing some artifacts over others is important to ensure that the protocol can always make progress and not be starved of network bandwidth for "less important" traffic.
This principle is well known from traditional networking and applies equally well to a blockchain system.

> ## Sparse Overlay Network Topology

Reviewer 2: I'd omit this part. When we recorded the video, we thought we might do that soon. Looking forward today, I don't think we will do anything like that any time in the foreseeable future.

An artifact pool has a *validated* and an *unvalidated* section, containing artifacts that have been validated by the upper protocol layer of the node or generated on the node, or containing artifacts that are still unvalidated, respectively.
Different pools are used for different kinds of artifacts.
Whenever the P2P layer receives an artifact from another node, it becomes part of an unvalidated pool.
P2P then requests its validation from components in a higher protocol layer, and once validated, the artifact is moved to the validated pool for this kind of artifact.
Artifacts generated by the protocol on the node are immediately placed into a validated pool, as being generated by the protocol locally makes them implicitly validated.

The implementation uses *adverts queues* to track received adverts, a *requested queue* to track which artifacts are currently being downloaded from peers, and a *receive check* data structure that tracks which artifacts have been downloaded already.
Those data structures help realize an efficient advert-based gossip protocol supporting artifact priorities.

> Each node maintains multiple parallel open connections to each of its peers to better utilize the available network bandwidth.

Reviewer 2: This is not true in the *current* implementation. It will become true hopefully in a few weeks.

> A connection can be throttled to avoid one component at a higher protocol layer monopolizing the nodes's available network bandwidth.

Reviewer 2: This is not implemented and there is no concrete plan to do that any time soon (there is a less-than-concrete plan to do that some day).
